### ISSUE TITLE
rst-sphinx causes strange behaviors

### ISSUE THREAD

**Issue by xuhdev** [flycheck/flycheck#953]
rst-sphinx keeps complaining "Unknown directive type" in my Sphinx project buffers. The way to reproduce it is as follows.

1. Run `sphinx-quickstart` to create a minimal Sphinx project. Use the default to answer all questions.
2. `emacs index.rst`.
3. Now everything looks fine. Move the cursor to the end of the buffer, insert a space, but DON'T save the buffer. All the complains show up. However, if the buffer is saved, all complains are gone.

I'm on Sphinx 1.4.1. Run the checking command on the command line does not show any error.

**Comment by lunaryorn** [comment]
Hello.  Thank you for reporting this issue, but I'm sorry to say that this is actually expected behaviour.

Due to the nature of Sphinx' builds we cannot run them on the fly—Sphinx always wants to build files, and only files within its project tree, and we have really no way to tell Sphinx to build a temporary file, or use standard input for a certain document.  Essentially we've no way to feed unsaved buffer contents to Sphinx.

So we can only run Sphinx on saved buffers.  To provide at least some last resort on the fly checking we run the standard `rst` syntax checker on unsaved buffers which will catch obvious syntax errors but unfortunately also introduces spurious errors because it doesn't know about Sphinx' extensions to ReStructuredText.

I see that this behaviour is confusing, but it's the best that we can get for Sphinx.  If you'd like to remove the spurious errors then disable `rst` and change `flycheck-check-syntax-automatically` to only check on save, e.g.

```elisp
(add-hook 'rst-mode-hook (lambda () (setq-local flycheck-disabled-checkers '(rst) flycheck-check-syntax-automatically '(save mode-enabled)))
```

**Comment by xuhdev** [comment]
@lunaryorn I can understand that rst-sphinx cannot check the unsaved version, but the fact that it gives some error message does not really make sense for me -- if the saved version on the actual file system is correct, why would it give errors then?

**Comment by xuhdev** [comment]
Oh I see, you meant when the buffer is unsaved, `rst` is used. If this is the case, shouldn't it be considered a bug? I don't see how it makes sense to use two different checkers for saved and unsaved buffers in general.

**Comment by lunaryorn** [comment]
@xuhdev _We_ do not consider it a bug.  As I said, it's expected, and even intentional behaviour.

YMMV, but what's the alternative?  

We can't run `rst-sphinx`, for two reasons: It's taking too long to run on every change, particularly on large docs, and it's just wrong.  After all, it'll never actually check contents of a modified buffer, so we'd be giving a wrong promise.

**Comment by xuhdev** [comment]
@lunaryorn  I meant, for a checker that is chosen and cannot be used to check modified buffers, we shouldn't switch the checker for modified buffers -- we should simply do not invoke any checkers. I think this is better, as there is no obvious incentive to run two different syntax checkers in different cases.

**Comment by lunaryorn** [comment]
@xuhdev  Why not?  I find that there's absolutely an incentive to check as much as possible, and I the reporting of `rst` still helpful while writing, because it gives me instant feedback on type errors or missing reference targets.

That said, it's surely a matter of taste, in my opinion, so we need to allow for both options, I guess. 

Within Flycheck's current implementation that comes down to keeping `rst` on by default and letting you disable it if you don't like it, because there's no feature within Flycheck to disable a checker by default except under certain circumstances (namely a plain rst file) while still allowing to fully enable the checker if one wishes.

**Comment by raxod502** [comment]
My solution:

```elisp
(defun radian-flycheck-maybe-disable-rst ()
  "If inside Sphinx project, disable the `rst' checker from `flycheck'.
This prevents it from signalling spurious errors."
  (with-eval-after-load 'flycheck
    (when (locate-dominating-file default-directory "conf.py")
      (make-local-variable 'flycheck-disabled-checkers)
      (push 'rst flycheck-disabled-checkers)
      (setq-local flycheck-check-syntax-automatically '(save mode-enabled)))))

(add-hook 'rst-mode-hook #'radian-flycheck-maybe-disable-rst)
```
